### PR TITLE
util: remove unused cpp-subprocess options

### PR DESCRIPTION
**Pull request by theStack**
The newly introduced cpp-subprocess library provides a good number of options for the `Popen` class:
https://github.com/bitcoin/bitcoin/blob/0de63b8b46eff5cda85b4950062703324ba65a80/src/util/subprocess.hpp#L1009-L1020
Some of them are either not fully implemented (`shell`, missing an implementation on Windows), implemented in an ugly way (e.g. using "Impoverished, meager, needy, truly needy version of type erasure" for `preexec_func` according to the author's own words) or simply unlikely to be ever needed for our external signer use-case (`defer_spawn`). Instead of maintaining incomplete and/or unneeded code, I'd suggest to get rid of it and only keep support for options if there is a strong reason for it.

### CODE REVIEW

**Comment by TheCharlatan** [comment]
Nice, Concept ACK.

**Comment by hebasto** [comment]
Concept ACK.

**Comment by hebasto** [comment]
> remove unused cpp-subprocess options

Are all remaining options used?

**Comment by laanwj** [comment]
Concept ACK, but could it be that we could need some of these?
Eg `close_fds` sounds like what's described in #22417

**Comment by laanwj** [comment]
> For example, setting environment variables seems like it could be useful, but the code is [quite complex for Windows](https://github.com/bitcoin/bitcoin/blob/07720b1cdd77399f32124641dbe1dd267eb0cf8b/src/util/subprocess.hpp#L309-L400).

Agree that that seems useful. I vaguely remember we even had a use-case for that once, but don't remember it clearly, we did give it up because it was too hard to do in windows. 

> Good question. Note that close_fds is currently not removed.

Oh, whoops, sorry, didn't notice that.

FWIW I think it's a valid choice to remove what we're not using and re-introduce it when we do, the code is out there there's little point in keeping unused code in the repository.

**Comment by hebasto** [comment]
> FWIW I think it's a valid choice to remove what we're not using and re-introduce it when we do, the code is out there there's little point in keeping unused code in the repository.

I agree.

**Review by hebasto — COMMENTED** [review]
Approach ACK 0f847336428153042dbf226caf52ecd43188f22e.

image :rocket: 

I have reviewed all commits except the last one (it's still in progress).

The CI run of the combined branch [ this PR + #29868 ] is available here: https://github.com/hebasto/bitcoin/actions/runs/8740663547/job/23984855961

File: src/util/subprocess.hpp [hunk 1]
```diff
@@ -1336,11 +1211,6 @@ inline void Popen::execute_process() noexcept(false)
 {
 #ifdef __USING_WINDOWS__
   void* environment_string_table_ptr = nullptr;
```
Comment (achow101) [inline src/util/subprocess.hpp:1213]: In 0f847336428153042dbf226caf52ecd43188f22e "remove unneeded environment option from cpp-subprocess"

This variable seems to also be unused now as it is never set to anything. It's only passed in as an argument to `CreateProcessW` which could take the nullptr directly.

**Review by hebasto — APPROVED** [review]
ACK 0f847336428153042dbf226caf52ecd43188f22e.

I'm happy to re-ACK if the https://github.com/bitcoin/bitcoin/pull/29865#discussion_r1571141464 will be addressed.

**Comment by luke-jr** [comment]
This seems likely to have a high maintenance cost. What's the benefit?

File: src/util/subprocess.hpp [hunk 2]
```diff
@@ -1549,7 +1256,7 @@ inline void Popen::execute_process() noexcept(false)
                             NULL,         // primary thread security attributes
                             TRUE,         // handles are inherited
                             creation_flags,	// creation flags
-                            environment_string_table_ptr,  // use provided environment
+                            NULL,         // use environment of the calling process
```
Comment (hebasto) [inline src/util/subprocess.hpp:1259]: nit: Maybe a wording more consistent with other comment around:
```suggestion
                            NULL,         // use parent's environment
```

**Review by hebasto — APPROVED** [review]
re-ACK 899be1a7643040c32510903dab06cea5ef7a5ffd.

**Review by hebasto — APPROVED** [review]
re-ACK 13adbf733f09c73c3cf0025d94c52f9cec5dba3b.

**Comment by achow101** [comment]
ACK 13adbf733f09c73c3cf0025d94c52f9cec5dba3b

**Comment by hebasto** [comment]
https://github.com/bitcoin/bitcoin/pull/29910 is the next one :)

**Pull request merged**